internal/controller/streams.go

```diff
@@ -363,76 +367,50 @@ func connectAndStreamLogs(logger *zap.Logger, sm *streamManager) error {
 }
 
 // Generic function to manage any stream with backoff and reconnection logic.
-func manageStream(logger *zap.Logger, connectAndStream func(*zap.Logger, *streamManager) error, sm *streamManager, done chan struct{}) {
+func manageStream(
+	logger *zap.Logger,
+	connectAndStream func(*zap.Logger, *streamManager) error,
+	sm *streamManager,
+	done chan struct{},
+) {
 	defer close(done)
-	const (
-		initialBackoff       = 1 * time.Second
-		maxBackoff           = 1 * time.Minute
-		maxJitterPct         = 0.20
-		resetPeriod          = 10 * time.Minute
-		severeErrorThreshold = 10 // Define what constitutes a severe error.
-	)
-
-	var (
-		backoff             = initialBackoff
-		consecutiveFailures = 0
-	)
-
-	resetTimer := time.NewTimer(resetPeriod)
-	sleepTimer := time.NewTimer(initialBackoff) // Start with an initial backoff interval
-	defer sleepTimer.Stop()
-	for {
-		select {
-		case <-resetTimer.C:
-			consecutiveFailures = 0
-			backoff = initialBackoff
-			resetTimer.Reset(resetPeriod)
-		case <-sleepTimer.C:
-			err := connectAndStream(logger, sm)
-			if err != nil {
-				if errors.Is(err, ErrStopRetries) {
-					logger.Info("Stopping retries for this stream as instructed.")
-					return
-				}
-				logger.Error("Failed to establish stream connection; will retry", zap.Error(err))
-				switch {
-				case consecutiveFailures == 0:
-					resetTimer.Reset(resetPeriod)
-				case consecutiveFailures >= severeErrorThreshold:
-					sleepTimer.Reset(resetPeriod)
-					<-resetTimer.C // Wait for reset timer to reset the failure count
-					consecutiveFailures = 0
-					backoff = initialBackoff
-					resetTimer.Reset(resetPeriod)
-					continue
-				}
-
-				consecutiveFailures++
-
-				sleep := jitterTime(backoff, maxJitterPct)
-				if sleep < initialBackoff {
-					sleep = initialBackoff
-				}
-				if sleep > maxBackoff {
-					sleep = maxBackoff
-				}
-
-				logger.Info("Sleeping before retrying connection", zap.Duration("backoff", sleep))
-				sleepTimer.Reset(sleep) // Reset sleep timer with calculated backoff delay
-
-				backoff *= 2
-				if backoff > maxBackoff {
-					backoff = maxBackoff
-				}
-			} else {
-				consecutiveFailures = 0
-				backoff = initialBackoff
-				// Reset sleep timer back to initial backoff interval
-				sleepTimer.Reset(initialBackoff)
-				resetTimer.Reset(resetPeriod)
-			}
-		}
+
+	lambda := func() error {
+		return connectAndStream(logger, sm)
+	}
+
+	// If a stream goes down, try to reconnect. With exponential backoff
+	lambdaWithBackoff := func() error {
+		return exponentialBackoff(backoffOpts{
+			InitialBackoff:       1 * time.Second,
+			MaxBackoff:           1 * time.Minute,
+			MaxJitterPct:         0.20,
+			SevereErrorThreshold: 10,
+			ExponentialFactor:    2.0,
+			Name:                 "connectToStreamWithBackoff",
+			logger:               logger,
```

Comment:
1. Why not give the name of the function that is called?
1. Instead of passing in the name as a parameter and then adding it explicitly in every log statement, just set it here in the logger's context only once.
```suggestion
			logger:               logger.With(zap.String("name", "connect_and_stream")),
```